### PR TITLE
[egs] icsi recipe

### PR DESCRIPTION
**Pull request by pswietojanski**
Hey,

I had parts of this recipe for some time, but never really had time to finish it properly, until recently. I added missing bits to make it fully run-able (given someone's access to ICSI data). One can build models for three conditions with some flexibility w.r.t what mics to use in sdm and mdm cases.

This PR includes results for GMM systems, at some point I will update this with mini-librispeech compatible chain models (but due to time constraints, this may take some time, so just committing baseline systems for now). Perhaps someone else will get to nnet3 baselines faster,  given data prep is there and reasonably tested.

Recipe is based on AMI s5b as much as possible. Let me know if there is something you would like to me to update, before you are happy with PR quality.

Best,
Pawel

### CODE REVIEW

**Comment by danpovey** [comment]
Great-- thanks!
How does it relate to the different versions of the AMI scripts?
There might be a couple of unnecessary files there e.g. topo.proto and plp.conf.

**Comment by pswietojanski** [comment]
conf/* is basically copied (except beamformer.cfg). The data preparation pipeline is entirely custom, to take into account all icsi data specifics. local/convert2stm.py and local/glm are the same (should I soft-link them?).

**Comment by danpovey** [comment]
I prefer copies to soft links.
Let's wait and see if anyone who is familiar with the AMI setup (Vimal?
Shinji?) can comment.

**Comment by pswietojanski** [comment]
Thanks. So the reason I decided for LDA distribution in the end is that it offers complete corpora (including all distant channels). This is something that seems to be missing from Edinburgh download (only ihm). And I assumed most universities have LDC-released ICSI data already in their repositories. I do mention http://groups.inf.ed.ac.uk/ami/icsi/ in the README though, so ppl are aware at least ihm should be available from there.

**Comment by pswietojanski** [comment]
Hey @danpovey so I had some time over the last weekend and added / trained chain models for each ICSI scenario (including ihm alignments for mdm and sdm setups). I guess this is ready to be merged at this point, I am not planning any further additions in the foreseeable future.

File: egs/icsi/README.txt [hunk 1]
```diff
@@ -0,0 +1,99 @@
+
+To train models, for an arbitrary mic, go to s5 run something like (after you
```
Comment (danpovey) [inline egs/icsi/README.txt]: can you please move the "instructional" type recipes to a comment at the top of the run.sh, and have this README.txt more about the data itself?  Please prominently say how many hours of train/test, what language, what type of speech (I assume meeting, with overlaps and multiple microphones), and what sampling rate.

Comment (danpovey) [reply egs/icsi/README.txt]: ... and also make clear prominently, at the top, that it's available from the LDC and mention the corpus numbers.
The idea is that people who are looking for setups with free data will know quickly not to look hdre.

**Comment by pswietojanski** [comment]
Thanks for the feedback, updated as requested. Fail of the check build of the latest update to this PR seems independent of it, I guess some infrastructure error on Travis side.

**Comment by jtrmal** [comment]
yes, I have even tried to restart the build and it failed again. I also
think it's infrastructure problem.
y.

**Comment by pswietojanski** [comment]
Should I do some dummy PR update to re-trigger the checks today, or one can do in some other way?

**Comment by jtrmal** [comment]
I have restarted it.
y.

**Comment by pswietojanski** [comment]
I just noticed you were not explicitly mentioned as the reviewers of this PR (thus perhaps overlooked this) @sw005320 @vimalmanohar Please could you have a look at your spare time? Cheers, Pawel

**Comment by sw005320** [comment]
Sure, I'll check some of them (possibly weekend).

File: egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+# Pawel: this one starts from the best tdnn model for ami, those are fairly similar
+# datasets so I would expect similar modelling tricks will work for both
+
+set -e -o pipefail
+# First the options that are passed through to run_ivector_common.sh
+# (some of which are also used in this script directly).
+stage=0
+mic=ihm
+nj=30
+min_seg_len=1.55
```
Comment (vimalmanohar) [inline egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh]: Recent scripts don't need min_seg_len and can instead provide multiple frames-per-chunk like 160,140,100 etc. The AMI recipe is also not updated with this change. But this is something that can be done in this recipe to reduce the many complexities involved in combing short segment feats.

File: egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+use_ihm_ali=false
+train_set=train
+gmm=tri3  # the gmm for the target data
+ihm_gmm=tri3  # the gmm for the IHM system (if --use-ihm-ali true).
+num_threads_ubm=3
+ivector_transform_type=pca
```
Comment (vimalmanohar) [inline egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh]: I think it is better to make the pca version as default and remove it as an option.

File: egs/icsi/s5/local/icsi_beamform.sh [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+# End configuration section
+echo "$0 $@"  # Print the command line for logging
+
+[ -f ./path.sh ] && . ./path.sh; # source the path.
+. parse_options.sh || exit 1;
+
+if [ $# != 3 ]; then
+   echo "Wrong #arguments ($#, expected 4)"
+   echo "Usage: steps/ami_beamform.sh [options] <num-mics> <ami-dir> <wav-out-dir>"
```
Comment (vimalmanohar) [inline egs/icsi/s5/local/icsi_beamform.sh]: ami -> icsi

File: egs/icsi/s5/local/nnet3/prepare_lores_feats.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  echo "$0: making MFCC features for speed-perturbed ${maybe_ihm}data"
+  steps/make_mfcc.sh --nj $nj \
+    --cmd "$train_cmd" data/${mic}/${train_set}${ihm_suffix}_sp
+  steps/compute_cmvn_stats.sh data/${mic}/${train_set}${ihm_suffix}_sp
+  echo "$0: fixing input data-dir to remove nonexistent features, in case some "
+  echo ".. speed-perturbed segments were too short."
+  utils/fix_data_dir.sh data/${mic}/${train_set}${ihm_suffix}_sp
+fi
+
+#we do not combine segments here by default, as have not done experiment if it helps
```
Comment (vimalmanohar) [inline egs/icsi/s5/local/nnet3/prepare_lores_feats.sh]: It seems like min_seg_len is not even used. It is better to remove this and try instead to provide multiple frame-chunks.

File: egs/icsi/s5/conf/beamformit.cfg [hunk 6]
```diff
@@ -0,0 +1,56 @@
+#BeamformIt sample configuration file for AMI data (http://groups.inf.ed.ac.uk/ami/download/)
```
Comment (sw005320) [inline egs/icsi/s5/conf/beamformit.cfg]: AMI -> ICSI
Also, remove the URL

File: egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+EOF
+
+  steps/nnet3/xconfig_to_configs.py --xconfig-file $dir/configs/network.xconfig --config-dir $dir/configs/
+fi
+
+if [ $stage -le 16 ]; then
+  if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d $dir/egs/storage ]; then
+    utils/create_split_dir.pl \
+     /export/b0{5,6,7,8}/$USER/kaldi-data/egs/ami-$(date +'%m_%d_%H_%M')/s5b/$dir/egs/storage $dir/egs/storage
```
Comment (sw005320) [inline egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh]: ami->icsi

File: egs/icsi/s5/local/icsi_agree_words_replace.pl [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/perl
+
+if (@ARGV != 3) {
+  print STDERR "Usage: ami_agree_words.pl <replace-file> <in-segments> <out-segments>\n";
```
Comment (sw005320) [inline egs/icsi/s5/local/icsi_agree_words_replace.pl]: ami->icsi

File: egs/icsi/s5/local/icsi_agree_words_split.pl [hunk 9]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/perl
+
+if (@ARGV != 3) {
+  print STDERR "Usage: ami_agree_words_split.pl <split-file> <in-segments> <out-segments>\n";
```
Comment (sw005320) [inline egs/icsi/s5/local/icsi_agree_words_split.pl]: ami->icsi

File: egs/icsi/s5/run.sh [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+nj=30 # number of parallel jobs,
+stage=0
+. utils/parse_options.sh
+
+base_mic=$(echo $mic | sed 's/[0-9]//g') # sdm, ihm or mdm
+nmics=$(echo $mic | sed 's/[a-z]//g') # e.g. 8 for mdm8.
+
+set -euo pipefail
+
+# Path where AMI gets downloaded (or where locally available):
```
Comment (sw005320) [inline egs/icsi/s5/run.sh]: These comments look like for AMI

File: egs/icsi/s5/run.sh [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+  done
+fi
+
+# monophone training
+if [ $stage -le 5 ]; then
+  # Full set 77h, reduced set 10.8h,
+  utils/subset_data_dir.sh data/$mic/train 15000 data/$mic/train_15k
+
+  steps/train_mono.sh --nj $nj --cmd "$train_cmd" \
+    data/$mic/train_15k data/lang exp/$mic/mono
```
Comment (sw005320) [inline egs/icsi/s5/run.sh]: It depends on the task, and amount of training data, but I usually observed that training with original noisy speech is better than that with enhanced speech (or using both), when we use DNN acoustic models. Using all microphone data is also good in terms of we'll increase the amount of speech data. The chime5 recipe is similar except for this part.

Comment (danpovey) [reply egs/icsi/s5/run.sh]: @sw005320, can you please clarify that comment above?
Were you asking for them to make some specific change in the recipe? 
Are you saying that the enhancement is not helpful for training but can still help for test?

Comment (sw005320) [reply egs/icsi/s5/run.sh]: > Are you saying that the enhancement is not helpful for training but can still help for test?
Right.
But this needs experiments, as it is not always true. This is just a comment.

**Comment by pswietojanski** [comment]
Thanks for the feedback, cleaned some top level bits (mostly typos) and removed variables that were not used anyway (but were still declared as overlooked this). At this point will not change a default chunk_width (from 150), as would need to re-train chain models for that, and do not have capacity to do so. I would expect this to be a less of an issue for ICSI (when compared to AMI), as there are fewer shorter segments overall (below 1.5s) but anyway, it is substantial so will have a look at this, when re-training chain models will be an option on my side.

**Comment by pswietojanski** [comment]
An update, re-ran all mic conditions with frames_per_eg set to 150,110,100. The results are updated accordingly.

**Comment by danpovey** [comment]
@pswietojanski, don't forget about this!  I think the changes I asked for were quite small.

**Comment by pswietojanski** [comment]
I will try to push another set of fixes this or next week (at latest). Thanks for the feedback Dan.

**Comment by danpovey** [comment]
@pswietojanski are you still intending to work on this?

**Comment by pswietojanski** [comment]
> @pswietojanski are you still intending to work on this?

Hi Dan. Apologies, I was changing jobs at that time, and somewhat overlooked and forgot about this. I guess this recipe is quite old already. I have some time next week (and spare resources), could you recommend which chain model is up to date so I can swap and rerun stuff so it is up to date? Is updating to mini_librispeech OK?

**Comment by jtrmal** [comment]
@aarora8 have a look

**Comment by pswietojanski** [comment]
> @aarora8 have a look

Hi @jtrmal and @aarora8 I commited fixes Dan asked for (removed dependency on irlstm, now all is done with srilm), and some other minor things. I rerun first stages of the recipe (data prep + gmm) to make sure these are comparable, and work properly.

1) There are still issues reported by CodeFactor in bash scripts (this hasn't been checked when original PR was made). Are these mandatory to resolve?
2) W.r.t chain, I am more than happy to merge the current models But if it's necessary, I may try to align these with what is currently SOTA model for AMI/s5b recipe.

Thanks both.

**Comment by aarora8** [comment]
Thanks, I will try to run it.

**Comment by danpovey** [comment]
Thanks,  guys!  Ashish, make a reminder to get back to us even if you don't test it, I can consider merging it anyway.

**Comment by aarora8** [comment]
ok.

**Comment by aarora8** [comment]
For the GMM-HMM setup, I am getting a similar dev/eval WER (35.9/26.1) as in the RESULTS file.

**Comment by pswietojanski** [comment]
I am also training ihm/mdm/sdm models to get more recent chain-based numbers comparable with AMI. I should commit these later today (so please do not merge at least until then).

**Comment by pswietojanski** [comment]
Hi, so I pushed the recipes / results for more up-to-date FTDNN models. I am not planning to work on this PR anymore, unless you spot something that needs fixes.

**Comment by danpovey** [comment]
Thanks!  Merged.

**Pull request merged**